library/modules/EventManager.cpp

```diff
@@ -491,31 +497,39 @@ static void manageJobStartedEvent(color_ostream& out) {
     // iterate event handler callbacks
     multimap<Plugin*, EventHandler> copy(handlers[EventType::JOB_STARTED].begin(), handlers[EventType::JOB_STARTED].end());
 
-    unordered_set<int32_t> newStartedJobs;
-
-    for (df::job_list_link* link = &df::global::world->jobs.list; link->next != nullptr; link = link->next) {
-        df::job* job = link->next->item;
-        if (!job || !Job::getWorker(job))
-            continue;
-
-        int32_t j_id = job->id;
-        newStartedJobs.emplace(j_id);
-        if (!startedJobs.count(j_id)) {
-            for (auto &[_,handle] : copy) {
-                DEBUG(log,out).print("calling handler for job started event\n");
-                run_handler(out, EventType::JOB_STARTED, handle, job);
+    std::vector<int32_t> newStartedJobs;
+    newStartedJobs.reserve(startedJobs.size());
+
+    for (const auto jobPtr : df::global::world->jobs.list) {
+        // posting_index of -1 implies a worker has been assigned to a new job.
+        if (jobPtr->posting_index == -1) {
+            auto jobId = jobPtr->id;
+            newStartedJobs.push_back(jobId);
+            /*
+             * The startedJobs set peaks at the number of work-eligible citizens.
+             * This set is small enough to fit comfortably in the CPU caches,
+             * e.g., 200 workers * 4 bytes (jobId) = 800 bytes,
+             * ensuring better memory locality and thus more efficiency than a hashmap
+             * where memory access tends to be all over the place.
+             */
+            if (!std::binary_search(startedJobs.begin(), startedJobs.end(), jobId)) {
+                for (auto &[_,handle] : copy) {
+                    DEBUG(log,out).print("calling handler for job started event\n");
+                    run_handler(out, EventType::JOB_STARTED, handle, jobPtr);
+                }
             }
         }
     }
-
-    startedJobs = newStartedJobs;
+    startedJobs = std::move(newStartedJobs);
 }
 
 //helper function for manageJobCompletedEvent
-//static int32_t getWorkerID(df::job* job) {
-//    auto ref = findRef(job->general_refs, general_ref_type::UNIT_WORKER);
-//    return ref ? ref->getID() : -1;
-//}
+#if 0
+static int32_t getWorkerID(df::job* job) {
+    auto ref = findRef(job->general_refs, general_ref_type::UNIT_WORKER);
+    return ref ? ref->getID() : -1;
+}
+#endif
```

Comment:
can we just get rid of this?

Reply:
Yes, it was already there, and used it myself, but there's other functions to use for that. Not needed.